### PR TITLE
Tensor changes in master graph and ring buffer

### PR DESCRIPTION
**Pull request by fiona-gladwin**
- Change all instances of Image to rocalTensor

**In Master Graph**

- Removed the internal batch size related params and computations
- Removed the output routine for video and making use of single loader routine for image and vide pipeline
- Remove variables associated with the above changes
- Remove output tensor allocation and deallocation (used for output copy)
- Remove copy_data API
- Introduced get_output_tensors API to directly return output tensor list

**In Ring Buffer**

- Remove master host buffers
- Modified host sub buffers to be used similar to device sub buffers

### CODE REVIEW

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 1]
```diff
@@ -320,30 +310,28 @@ MasterGraph::create_tensor(const rocalTensorInfo &info, bool is_output) {
...
-MasterGraph::set_output(Image* output_image)
+MasterGraph::set_output(rocalTensor* output_tensor)
 {
-    if(output_image->is_handle_set() == false)
+    if(output_tensor->is_handle_set() == false)
     {
-        if (output_image->create_from_handle(_context) != 0)
-                THROW("Cannot create the image from handle")
-        _output_images.push_back(output_image);
+        if (output_tensor->create_from_handle(_context) != 0)
```
Comment (rrawther) [inline rocAL/rocAL/source/pipeline/master_graph.cpp:324]: tensors are created in both set_output() and create_tensor() call. We need to make sure, it is created only once for each output.

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 2]
```diff
@@ -535,428 +428,29 @@ MasterGraph::mem_type()
...
-    else {
-#endif
-        // get_host_master_read_buffer is blocking if _ring_buffer is empty, and blocks this thread till internal processing thread process a new batch and store in the _ring_buffer
-        memcpy(out_ptr, _ring_buffer.get_host_master_read_buffer(), size * _output_images.size());
-#if ENABLE_OPENCL || ENABLE_HIP
-    }
-#endif    
-    _convert_time.end();
-    return Status::OK;
+    std::vector<void*> output_ptr = _ring_buffer.get_read_buffers();
```
Comment (rrawther) [inline rocAL/rocAL/source/pipeline/master_graph.cpp]: Is this call doing a copy or returning reference to a vector? Can you please use auto call like the image pipeline here?

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 3]
```diff
@@ -535,428 +428,29 @@ MasterGraph::mem_type()
...
+    for(unsigned i = 0; i < _internal_tensor_list.size(); i++)
+        _output_tensor_list[i]->set_mem_handle(output_ptr[i]);
+    
+    return &_output_tensor_list;
```
Comment (rrawther) [inline rocAL/rocAL/source/pipeline/master_graph.cpp:445]: In case of framework integration: how does it copy to framework tensor?

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 4]
```diff
@@ -1062,152 +542,20 @@ void MasterGraph::output_routine()
...
             _ring_buffer.push(); // Image data and metadata is now stored in output the ring_buffer, increases it's level by 1
         }
-        _process_time.end();
-
-    }
-    catch (const std::exception &e)
-    {
-        ERR("Exception thrown in the process routine: " + STR(e.what()) + STR("\n"));
-        _processing = false;
-        _ring_buffer.release_all_blocked_calls();
```
Comment (rrawther) [inline rocAL/rocAL/source/pipeline/master_graph.cpp:1072]: why is this exception handling code removed?

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 5]
```diff
@@ -1422,64 +770,6 @@ size_t MasterGraph::bounding_box_batch_count(int *buf, pMetaDataBatch meta_data_
     return size;
 }
 
-
-size_t MasterGraph::compute_optimum_internal_batch_size(size_t user_batch_size, RocalAffinity affinity)
```
Comment (rrawther) [inline rocAL/rocAL/source/pipeline/master_graph.cpp:1426]: We need similar function to calculate OMP number of threads where user doesn't pass that

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 6]
```diff
@@ -1493,163 +783,6 @@ bool MasterGraph::no_more_processed_data()
     return (_output_routine_finished_processing && _ring_buffer.empty());
 }
 
-MasterGraph::Status
-MasterGraph::copy_out_tensor_planar(void *out_ptr, RocalTensorFormat format, float multiplier0, float multiplier1,
```
Comment (rrawther) [inline rocAL/rocAL/source/pipeline/master_graph.cpp:1497]: this was required for supporting planar output from the pipeline (like for CFar10 data set). Is this handled in tensor pipeline?

File: rocAL/rocAL/source/pipeline/ring_buffer.cpp [hunk 7]
```diff
@@ -159,7 +151,7 @@ void RingBuffer::init(RocalMemType mem_type, void *devres, unsigned sub_buffer_s
 
         }
     }
-   else
+    else
```
Comment (rrawther) [inline rocAL/rocAL/source/pipeline/ring_buffer.cpp:154]: make sure the code works for all 3 backends (ENABLE_HIP, ENABLE_OPENCL, CPU)

File: rocAL/rocAL/source/pipeline/tensor.cpp [hunk 8]
```diff
@@ -316,7 +316,7 @@ unsigned rocalTensor::copy_data(hipStream_t stream, void *host_memory, bool sync
 #endif
 
 unsigned rocalTensor::copy_data(void *user_buffer) {
-    if (_info._type != rocalTensorInfo::Type::HANDLE) return 0;
```
Comment (rrawther) [inline rocAL/rocAL/source/pipeline/tensor.cpp:319]: why is this changed? Are they both have the same meaning

**Review by rrawther — CHANGES_REQUESTED** [review]
@fiona-gladwin : please address review comments

File: rocAL/rocAL/source/pipeline/master_graph.cpp (hunk 1 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/pipeline/master_graph.cpp:324]: We are checking if the handle is set and then create the output tensors here to ensure we create the output only once

File: rocAL/rocAL/source/pipeline/master_graph.cpp (hunk 3 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/pipeline/master_graph.cpp:445]: We would return the tensor to user and in the iterator we call a `tensor.copy_data(user_buffer)` where we pass a user buffer and the contents of tensor is copied to user buffer. This API is added as a member function of rocalTensor class in [tensor.cpp](https://github.com/LakshmiKumar23/MIVisionX/blob/a321eeb158deb9789b36104637b5538dfe7a985b/rocAL/rocAL/source/pipeline/tensor.cpp#L318)

File: rocAL/rocAL/source/pipeline/master_graph.cpp (hunk 4 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/pipeline/master_graph.cpp:1072]: We have removed the output routine of the video, so it is showing this as removed but the code is present for the same

File: rocAL/rocAL/source/pipeline/master_graph.cpp (hunk 5 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/pipeline/master_graph.cpp:1426]: We are working on the changes required to pass the threads from the user. We will add it in subsequent commits

File: rocAL/rocAL/source/pipeline/tensor.cpp (hunk 8 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/pipeline/tensor.cpp:319]: In our previous case we would associate all the image related rocalTensors with a vx_tensor and we set the type as handle for the output tensors where we create vx_tensors with handle.
But with our current changes we are not creating the output tensor (which is to be returned to the user) as a vx_tensor hence handle is not set, so we could not copy the data. Hence I have changed the condition to check if it is a null pointer.

File: rocAL/rocAL/source/pipeline/master_graph.cpp (hunk 2 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/pipeline/master_graph.cpp]: Changed this to auto. This call would return a copy of the vector of pointers

File: rocAL/rocAL/source/pipeline/master_graph.cpp (hunk 6 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/pipeline/master_graph.cpp:1497]: Yes this is handled in tensor pipeline, some of our trainings require planar output i.e NCHW we are able to return a planar output from rocAL and normalization would happen in RPP.

**Comment by fiona-gladwin** [comment]
Hi @rrawther and @LakshmiKumar23 I have addressed all the PR comments.

**Comment by LakshmiKumar23** [comment]
> Hi @rrawther and @LakshmiKumar23 I have addressed all the PR comments.

@rrawther can I merge this PR?

File: rocAL/rocAL/source/pipeline/master_graph.cpp (hunk 3 above)
Comment (rrawther) [reply rocAL/rocAL/source/pipeline/master_graph.cpp:445]: Hi Fionna, In that case we need to make copy_data() uses optimized code to get the best performance

**Review by rrawther — APPROVED** [review]
I see that many functions for copy() are removed from this.
We need to make sure we handle all those cases in the tensor pipeline
